### PR TITLE
Made `__repr__` for `Variable.data` consistent with that of `VariableSlice.data`

### LINKED ISSUE #107: `__repr__` of `Variable.data` and `VariableSlice.data` should be consistent
For `VariableSlice.data`, `__repr__` is defined to give a formatting similar to what you get for numpy arrays (implemented in `declare_VariableView`). For `Variable.data` this is missing.

Implement the same `__repr__` in `declare_span`.

### PR DESCRIPTION
**Pull request by nvaytet**
Copied the `__repr__` from `declare_VariableView` to `declare_span` and moved the `print` function
Fixes #107

### CODE REVIEW

File: exports/dataset.cpp [hunk 1]
```diff
@@ -781,6 +803,11 @@ PYBIND11_MODULE(dataset, m) {
              return slice;
            },
            py::keep_alive<0, 1>())
+      .def("__setitem__",
+           [](VariableSlice &self, const std::tuple<Dim, py::slice> &index,
+              const VariableSlice &other) {
+             detail::pySlice(self, index).assign(other);
+           })
```
Comment (igudich) [inline exports/dataset.cpp]: Would be good to add __setitem__ for:
1. VariableSlice &self, const std::tuple<Dim, **gsl::index**> &index, const VariableSlice &other
2. Variable &self, const std::tuple<Dim, **gsl::index**> &index, const VariableSlice &other
then.

File: exports/dataset.cpp [hunk 2]
```diff
@@ -56,7 +70,24 @@ template <class T> void declare_span(py::module &m, const std::string &suffix) {
...
+         const gsl::index size = self.size();
+         if (size == 0)
+           return std::string("[]");
+         std::string s = "[";
+         for (gsl::index i = 0; i < self.size(); ++i) {
+           if (i == 4 && size > 8) {
+             s += "..., ";
+             i = size - 4;
+           }
+           s += print(self[i]);
```
Comment (OwenArnold) [inline exports/dataset.cpp]: `print`? I don't think this function is doing any input/output. Maybe `to_string` is better and more consistent.

File: exports/dataset.cpp [hunk 3]
```diff
@@ -56,7 +70,24 @@ template <class T> void declare_span(py::module &m, const std::string &suffix) {
       .def("__len__", &gsl::span<T>::size)
       .def("__iter__", [](const gsl::span<T> &self) {
         return py::make_iterator(self.begin(), self.end());
-      });
+      })
+      .def("__repr__",
+       [](const gsl::span<T> &self) {
```
Comment (OwenArnold) [inline exports/dataset.cpp]: You mention this being the same representation is being used elsewhere. Is it worth making the code common?

File: exports/dataset.cpp (hunk 2 above)
Comment (OwenArnold) [inline exports/dataset.cpp]: This doesn't actually print (no IO), maybe `to_string` would be a better name?

Comment (igudich) [reply exports/dataset.cpp]: @OwenArnold, print calls to_string with the same signature, so it wouldn't work, although I agree print is not perfect here.

File: exports/dataset.cpp (hunk 1 above)
Comment (nvaytet) [reply exports/dataset.cpp]: did you mean i should do this as part of this PR or part of #152 ?

File: exports/dataset.cpp (hunk 2 above)
Comment (nvaytet) [reply exports/dataset.cpp]: something inside the `print()` function is already called `to_string`. Any ideas for another name?

File: exports/dataset.cpp (hunk 1 above)
Comment (igudich) [reply exports/dataset.cpp]: @nvaytet, possibly, but if it has been already done for slices (ranges) here, a good idea is to have this also for indexes.

Comment (nvaytet) [reply exports/dataset.cpp]: yes it is a good idea, I added it in #152

File: exports/dataset.cpp (hunk 2 above)
Comment (igudich) [reply exports/dataset.cpp]: I have 2 suggestions:
1.
``` c++
namespace tostr {
using namespace std::to_string;
using namespace dataset::to_string;
template <class T>
std::string impl(const T& x) { return to_string(x); }
} // tostr

namespace detail {
template <class T> std::string to_string(const T &item) {
  if constexpr (std::is_same_v<T, std::string>)
    return {'"' + item + "\", "};
  else if constexpr (std::is_same_v<T, Eigen::Vector3d>)
    return {"(Eigen::Vector3d), "};
  else if constexpr (std::is_same_v<T,
                                    boost::container::small_vector<double, 8>>)
    return {"(vector), "};
  else
    return tostr::impl(item) + ", ";
}
} // detail
```

2.

```c++
emplate<typename T> struct HasStdToString{
private:
    static void detect(...);
    template<typename U> static decltype(std::to_string(std::declval<U>())) detect(const U&);
public:
    static constexpr bool value = !std::is_same<void, decltype(detect(std::declval<T>()))>::value;  // 
};

template <class T> std::string to_string(const T &item) {
  if constexpr (std::is_same_v<T, std::string>)
    return {'"' + item + "\", "};
  else if constexpr (std::is_same_v<T, Eigen::Vector3d>)
    return {"(Eigen::Vector3d), "};
  else if constexpr (std::is_same_v<T,
                                    boost::container::small_vector<double, 8>>)
    return {"(vector), "};
  else if constexpr (HasStdToString)
    return std::to_string(item) + ", ";
  else
    return dataset::to_string(item) + ", "';
}
```

**Pull request merged**